libyara/modules/macho.c

```diff
@@ -0,0 +1,1225 @@
+
+    // Segments and nested sections
+
+    declare_integer("number_of_segments");
+
+    begin_struct_array("segments");
+      declare_string("segname");
+      declare_integer("vmaddr");
+      declare_integer("vmsize");
+      declare_integer("fileoff");
+      declare_integer("fsize");
+      declare_integer("maxprot");
+      declare_integer("initprot");
+      declare_integer("nsects");
+      declare_integer("flags");
+      begin_struct_array("sections");
+        declare_string("sectname");
+        declare_string("segname");
+        declare_integer("addr");
+        declare_integer("size");
+        declare_integer("offset");
+        declare_integer("align");
+        declare_integer("reloff");
+        declare_integer("nreloc");
+        declare_integer("flags");
+        declare_integer("reserved1");
+        declare_integer("reserved2");
+        declare_integer("reserved3");
+      end_struct_array("sections");
+    end_struct_array("segments");
+
+    // Entry point and stack size
+
+    declare_integer("entry_point");
+    declare_integer("stack_size");
+
+  end_struct_array("file");
+
+  // Mach-O fat binary helper functions
+
+  declare_function("file_index", "i", "i", file_index_type);
+  declare_function("file_index", "ii", "i", file_index_subtype);
+  declare_function("ep_for_arch", "i", "i", entry_point_for_arch_type);
+  declare_function("ep_for_arch", "ii", "i", entry_point_for_arch_subtype);
+
+end_declarations;
+
+int module_initialize(
+    YR_MODULE* module)
+{
+  return ERROR_SUCCESS;
+}
+
+int module_finalize(
+    YR_MODULE* module)
+{
+  return ERROR_SUCCESS;
+}
+
+int module_load(
+    YR_SCAN_CONTEXT* context,
+    YR_OBJECT* module_object,
+    void* module_data,
+    size_t module_data_size)
+{
+  YR_MEMORY_BLOCK* block;
+  YR_MEMORY_BLOCK_ITERATOR* iterator = context->iterator;
+
+  // Prepare storage.
+  FAT_DATA* st = (FAT_DATA*)yr_malloc(sizeof(FAT_DATA));
+
+  if (!st)
+    return ERROR_INSUFFICIENT_MEMORY;
+
+  module_object->data = st;
+  macho_init_fat_data_storage(st);
+
+  foreach_memory_block(iterator, block)
+  {
+    void* block_data = block->fetch_data(block);
+    if (block_data == NULL)
+      continue;
+
+    uint32_t* magic = (uint32_t*)block_data;
+    // Parse classic Mach-O file.
+    if (macho_is_file_block(magic))
+    {
+      macho_parse_file(block_data, block->size, module_object, context);
+
+      // In case of classic Mach-O files, everything will be in the first block
+      // so program does not have to check rest of the blocks.
+
+      break;
+    }
+
+    // Parse Mach-O fat binary.
+    if (macho_is_fat_file_block(magic))
+    {
+      macho_parse_fat_file(block_data, block->size, module_object, context);
+
+      // In case of fat Mach-O files, some necessary data may be out of block.
+
+      st->parsed_size += block->size;
+      continue;
+    }
+
+    // Parse previously out of block files.
+    for (size_t i = 0; i < st->count; i++)
```

Comment:
This is not actually necessary. You can safely assume that a Mach-O file will come in a single block, no matter how big it is. Multiple blocks are mainly for scanning process memory, where each block represents memory region.